datadog-setup.php

```diff
@@ -47,23 +71,353 @@ function print_help()
 
 EOD;
 }
 
+// phpcs:disable PSR1.Classes.ClassDeclaration.MissingNamespace
+class IniRecord
+{
+    /** @var string */
+    public $setting;
+    /** @var string */
+    public $currentValue;
+    /** @var string */
+    public $defaultValue;
+    /** @var string */
+    public $iniFile;
+}
+
+/**
+ * @param array $options
+ * @return Iterator<string, IniRecord>
+ */
+function config_list(array $options)
+{
+    $iniSettings = get_ini_settings('', '', '');
+
+    // The first 3 are 'extension' type of settings.
+    $iniSettings = array_slice($iniSettings, CMD_CONFIG_NUM_SHIFT);
+
+    // Build an index by unique names for filtering.
+    $indexByName = array_column($iniSettings, null, 'name');
+
+    foreach (require_binaries_or_exit($options) as $command => $fullPath) {
+        $binaryForLog = ($command === $fullPath) ? $fullPath : "$command ($fullPath)";
+        echo "Datadog configuration for binary: $binaryForLog", PHP_EOL;
+
+        $iniFilePaths = find_all_ini_files(ini_values($fullPath));
+
+        foreach ($iniFilePaths as $iniFilePath) {
+            $iniFileSettings = parse_ini_file($iniFilePath, false, INI_SCANNER_RAW);
+            $settings = array_intersect_key($iniFileSettings, $indexByName);
+
+            foreach ($settings as $iniFileSetting => $currentValue) {
+                $iniSetting = $indexByName[$iniFileSetting];
+                $record = new IniRecord();
+                $record->setting = $iniFileSetting;
+                $record->currentValue = $currentValue;
+                $record->defaultValue = $iniSetting['default'];
+                $record->iniFile = $iniFilePath;
+                yield $record->setting => $record;
+            }
+        }
+    }
+}
+
+/**
+ * This function will print out all Datadog specific PHP INI settings. The list
+ * of Datadog specific settings is retrieved by calling `get_ini_settings()`. It
+ * will also show the default and INI file this setting was found.
+ * The output will be grouped by PHP binary, example:
+ *
+ * $ php datadog-setup.php config list --php-bin all
+ * Searching for available php binaries, this operation might take a while.
+ * Datadog configuration for binary: php (/opt/php/8.2/bin/php)
+ * datadog.profiling.enabled = On ; default: 1, INI file: /opt/php/etc/conf.d/98-ddtrace.ini
+ * datadog.profiling.experimental_allocation_enabled = On ; default: 1, INI file: /opt/php/etc/conf.d/98-ddtrace.ini
+ *
+ * @see get_ini_settings
+ * @return void
+ */
+function cmd_config_list(array $options)
+{
+    foreach (config_list($options) as $record) {
+        echo "$record->setting = $record->currentValue ; default: $record->defaultValue, INI file: $record->iniFile\n";
+    }
+}
+
+/**
+ * This function will print the specified PHP INI settings. It only prints
+ * information for Datadog's own settings; it can't be used to parse opache,
+ * xdebug, etc.
+ * The output will be grouped by PHP binary, example:
```

Comment:
It's grouped by INI, really, not saying which binary (which is fine), right?

Reply:
There is an echo in `config_list` which prints out the binary. The separation isn't clean as it probably ought to be. I wrote that when unifying config get and config list a bit.

Reply:
Oh. Bit awkward with the yield & echo mix :-D

Reply:
I'm about to check this, as with multiple binaries it gives
```bash
php datadog-setup.php config get -d datadog.profiling.enabled --php-bin /opt/php/8.2/bin/php --php-bin /opt/php/8.1/bin/php
Datadog configuration for binary: /opt/php/8.2/bin/php
Datadog configuration for binary: /opt/php/8.1/bin/php
datadog.profiling.enabled = On ; default: 1, INI file: /opt/php/8.2/etc/conf.d/98-ddtrace.ini
```
where it should be
```bash
php datadog-setup.php config get -d datadog.profiling.enabled --php-bin /opt/php/8.2/bin/php --php-bin /opt/php/8.1/bin/php
Datadog configuration for binary: /opt/php/8.2/bin/php
datadog.profiling.enabled = On ; default: 1, INI file: /opt/php/8.2/etc/conf.d/98-ddtrace.ini
Datadog configuration for binary: /opt/php/8.1/bin/php
```

Reply:
I moved the PHP binary to the `IniRecord` class, so that it can be printed out with the record itself.

Reply:
The output should now look like this:
```bash
$ php datadog-setup.php config get -d datadog.profiling.enabled --php-bin /opt/php/8.2/bin/php --php-bin /opt/php/8.1/bin/php
datadog.profiling.enabled = On; default: 1, binary: /opt/php/8.2/bin/php, INI file: /opt/php/8.2/etc/conf.d/98-ddtrace.ini
datadog.profiling.enabled = On; default: 1, binary: /opt/php/8.1/bin/php, INI file: /opt/php/8.2/etc/conf.d/98-ddtrace.ini
```